Test Row Group Skipping in SelectiveColumnReader

oerling opened the pull request:
Adds a test with column values that allow row group (stride)
skipping. Fixes seekToRowGroup for struct reader for a column with
nulls.

zzhao0 commented:
Line 3876, should pass true for loadIndex. However, that basically means we always pass true to load the index. We should only do that if we know column is projected or have filter on it.

oerling commented:
It turns out structs have an index stream for the nulls, just as they should. The problem is that the struct reader did not pass yje loadIndex argument.

 

loadIndex should be true if there is, anywhere in the reader tree, something that can skip a row group. This is not known at construction time because filters can be added by pushing down joins or such. So the index stream needs to always get retrieved from the streams. It needs to be decoded/loaded only on first seek to row group.

 

Just retrieving the stream from the StripeStreams is not expensive. The alternative is to mutate the whole reader tree on first pushdown, which is probably more trouble than worth.

zzhao0 commented on velox/dwio/dwrf/reader/SelectiveColumnReader.cpp:
```diff
@@ -74,27 +74,30 @@ SelectiveColumnReader::SelectiveColumnReader(
-  ensureRowGroupIndex();
-  auto filter = scanSpec_->filter();
-  if (!index_ || !filter) {
+  if ((!index_ && !indexStream_) || !scanSpec_->filter()) {
```
is !index_ && !indexStream_ a valid case? Looks to me we will always have one of them

oerling commented:
It has been seen that the top level struct reader has neither index_ nor indexStream_. It does not strictly speaking need and index because it has no state apart the state of its children.

I updated the diff, the asan failure in fbcode should be fixed as well as lint.

 

I changed setting the row group a little so that for top level readers, they do the seek on first read and not at the time the top level struct is given a row group.., Like this sparsely read readers do not read index or seek unless they are actually read.
